### PR TITLE
add query recording options

### LINKED ISSUE #140: Add dynamic recording
### Housekeeping

- [X] I am a maintainer of dbt-common

### Short description

We capture too much data with every run.  Add the ability to dynamically capture different data.

### Acceptance criteria

- we can capture data by record type

### Suggested Tests

test the different possible record types

### Impact to Other Teams

adapters if any changes are required for them

### Will backports be required?

no

### Context

Instead of capturing everything for every run, we should be able to specify what data we want to capture.  

Exactly what options we want to offer will be part of this ticket.  It should start with the minimal requirements (the data we will need to compare the SQL from one run with another run via the diff functionality

Add DBT_RECORD=True to your dbt command (ie. DBT_RECORD=True dbt run).  This generates a recording.json in the current working dir.

Options

This can be specified in the [RecorderMode](https://github.com/dbt-labs/dbt-common/blob/e1c192e41c5aaa0377beaaf280980eeb49b787a0/dbt_common/record.py#L59).

Use an env var in addition to DBT_RECORD/DBT_REPLAY to specify what we want to record

### PR DESCRIPTION
**Pull request by emmyoop**
resolves #140 / MNTL-268

### Description

Add option to record by record type.

### Checklist

- [ ] I have read [the contributing guide](https://github.com/dbt-labs/dbt-common/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [ ] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [ ] I have run this code in development and it appears to resolve the stated issue
- [ ] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [ ] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-common/blob/main/CONTRIBUTING.md#Adding-CHANGELOG-Entry)

### CODE REVIEW

**Review by peterallenwebb — CHANGES_REQUESTED** [review]

File: dbt_common/record.py [hunk 1]
```diff
@@ -147,17 +152,46 @@ def print_diffs(self) -> None:
 
 
 def get_record_mode_from_env() -> Optional[RecorderMode]:
```
Comment (ChenyuLInx) [inline dbt_common/record.py:154]: Not a blocker for the current PR but a general observation, I think we are basically implementing something similar to pydantic settings functionalities here. We should consider consolidating all env readings(other than the ones that are also flags in click commands) to use it.

File: tests/unit/test_record.py [hunk 2]
```diff
@@ -76,6 +136,10 @@ def test_func(a: int, b: str, c: Optional[str] = None) -> str:
 
     finally:
         if prev is None:
-            os.environ.pop("DBT_RECORD", None)
+            os.environ.pop("DBT_RECORDER_MODE", None)
```
Comment (ChenyuLInx) [inline tests/unit/test_record.py:139]: Setting up the env var and removing it might be easier to read if we write a fixture with yield in the middle?

**Review by ChenyuLInx — COMMENTED** [review]
This does not resolve #140  right? It just add the option part?

**Comment by emmyoop** [comment]
@ChenyuLInx the scope of #140 changed while working on this ticket.  I've updated it to reflect that we're just going to filter by record type.  So this will close #140.

File: dbt_common/record.py (hunk 1 above)
Comment (emmyoop) [reply dbt_common/record.py:154]: https://github.com/dbt-labs/dbt-core/issues/10100 addresses this.  Going to leave this as is for now.

**Review by peterallenwebb — APPROVED** [review]

File: dbt_common/record.py (hunk 1 above)
Comment (ChenyuLInx) [reply dbt_common/record.py:154]: Sounds good to me!

**Pull request merged**